### DISCUSSION TITLE
MCU change in YH-002A

### DISCUSSION THREAD

**Discussion by RichardGomer** [arendst/Tasmota#15153] [Ideas]
I've recently purchased some new YH-002A blind controllers. They're not working with Tasmota, and I wanted to share my findings in case anyone can shed some light on it, or in case anyone else has seen something similar.

I already have some older variants of the device, and I note that the PCB has changed in the new ones. Shown below: old device at the top, and a new one at the bottom.

![image](https://user-images.githubusercontent.com/2702099/158476267-0f1ebf99-c218-4a5e-80cc-c9820e0e85b3.png)

The new device does not respond to any of the TuyaSend<x> commands, on any DpId that I can find. Nor to the TuyaSend0/8 commands that should report status information.

I've attached a serial sniffer to a device running the stock Tuya firmware and captured the commands that the ESP sends to the MCU.

Here's the command sent by the Tuya firmware to move the blind to 30%:
`55 aa 00 06 00 08 02 02 00 00 00 00 1e 33`

And to 79%
`55 aa 00 06 00 08 02 02 00 00 00 00 4f 64`

I note that those commands differ very slightly to those sent by Tasmota, and seem to differ also from the Tuya serial protocol documentation. Up to `02 02` they're as expected - DpId = 2, and type = 2. 

BUT, there 'should' be two bytes following the type which indicate that payload length (in this case 4 bytes, so `00 04`) and that is indeed what Tasmota sends if I try the corresponding TuyaSend2 command. On the stock firmware it's as if the second byte of the length is just missing; the four bytes of the payload are all there, preceded just by a single 0 byte. (The very last byte is the checksum).

For comparison, TuyaSend 2,46 sends the following (according to the web console): `55 aa 00 06 00 08 02 02 00 04 00 00 00 2e 43`

I CAN see messages being received from the MCU when I operate the device using the hardware buttons, so receiving seems to work!

```
22:46:33.878 {"TuyaReceived":{"Data":"55AA0307000803020004000000223C","Cmnd":7,"CmndData":"0302000400000022","DpType2Id3":34,"3":{"DpId":3,"DpIdType":2,"DpIdData":"00000022"}}}
22:46:33.881 TYA: fnId=0 is set for dpId=3
22:46:33.883 TYA: RX value 34 from dpId 3 
22:46:33.891 {"TuyaReceived":{"Data":"55AA0307000507040001001A","Cmnd":7,"CmndData":"0704000100","DpType4Id7":0,"7":{"DpId":7,"DpIdType":4,"DpIdData":"00"}}}
22:46:33.893 TYA: fnId=0 is set for dpId=7
22:46:35.787 {"TuyaReceived":{"Data":"55AA0307000803020004000000031D","Cmnd":7,"CmndData":"0302000400000003","DpType2Id3":3,"3":{"DpId":3,"DpIdType":2,"DpIdData":"00000003"}}}
22:46:35.791 TYA: fnId=0 is set for dpId=3
22:46:35.793 TYA: RX value 3 from dpId 3
```

I've tried fiddling around with SerialSend/SSerialSend to see if I can get the MCU to respond to raw commands without success, so sadly can't verify if the single missing byte is the full issue.

Any ideas on what to debug next, or anyone seen anything similar?

**Comment by SuttoNZ247** [comment]
Hi Richard, 

Nice work with this. I am struggling to get the newer YH002 to communicate with MCU. did you have any more success with this? My symptoms are exactly the same as you have documented.

**Reply by RichardGomer** [reply]
I'm afraid they're still sat in pieces on my desk as they have been since March! I haven't yet mustered the enthusiasm to have another go, but with winter coming (and a desire to close the blinds automatically to reduce the heating bill) I do mean to try again. If I can make any progress I'll post it here :)

**Comment by pauly2805** [comment]
FYI:
https://github.com/esphome/issues/issues/3081
